### ISSUE TITLE
instances of twisted.web2.dav.element.base.WebDAVElement are not really hashable but are used as if they were

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2960]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2960|
|Type|defect|
|Created|2007-12-20 01:03:24Z|
|Branch|https://github.com/twisted/twisted/tree/dav-hash-2960|

This is visible when running the tests with current python trunk@HEAD which apparently fixes an oversight which allowed new-style classes to have a default hash implementation even when they override !__eq!__ and !__ne!__.

```
[ERROR]: twisted.web2.dav.test.test_mkcol

Traceback (most recent call last):
  File "/tmp/Twisted/twisted/trial/runner.py", line 591, in loadPackage
    module = modinfo.load()
  File "/tmp/Twisted/twisted/python/modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/tmp/Twisted/twisted/python/modules.py", line 618, in moduleLoader
    return self._moduleLoader(modname)
  File "/tmp/Twisted/twisted/python/reflect.py", line 361, in namedAny
    topLevelPackage = __import__(trialname)
  File "/tmp/Twisted/twisted/web2/dav/test/test_mkcol.py", line 30, in <module>
    from twisted.web2.dav.fileop import rmdir
  File "/tmp/Twisted/twisted/web2/dav/fileop.py", line 51, in <module>
    from twisted.web2.dav.http import ResponseQueue, statusForFailure
  File "/tmp/Twisted/twisted/web2/dav/http.py", line 45, in <module>
    from twisted.web2.dav import davxml
  File "/tmp/Twisted/twisted/web2/dav/davxml.py", line 46, in <module>
    from twisted.web2.dav.element.rfc2518 import *
  File "/tmp/Twisted/twisted/web2/dav/element/rfc2518.py", line 162, in <module>
    LockScope.exclusive = LockScope(Exclusive())
  File "/tmp/Twisted/twisted/web2/dav/element/base.py", line 326, in __new__
    child: WebDAVElement.__new__(clazz, children)
exceptions.TypeError: unhashable type: 'Exclusive'
```

<details><summary>Searchable metadata</summary>

```
trac-id__2960 2960
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__Python_2_6 Python-2.6
branch__branches_dav_hash_2960 branches/dav-hash-2960
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2_dav web2.dav
keywords__ 
time__1198112604000000 1198112604000000
changetime__1199914473000000 1199914473000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#2961](https://github.com/twisted/twisted/issues/2961) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22232]) Branching to 'dav-hash-2960'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22233]) Add an hash method to empty element.

Refs [#2960](https://github.com/twisted/twisted/issues/2960)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

This is ready to review. The only element which needed a hash was the empty element, and it's trivially hashable because it doesn't have child or attribute.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

What's the story behind all the changes aside from the new ```__hash__``` method?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

The ```__new__``` methods called ```WebDAVElement.__new__``` with arguments, whereas it's not supported (```object.__new__``` doesn't take extra arguments) and not necessary.

Python 2.6 starts to print a warning about that, so I removed the arguments.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @therve|
|-|-|

Okay, that makes sense.  There doesn't seem to be any real test coverage for this code, however the branch makes the module importable on python trunk@HEAD and doesn't break any of the ```twisted.web2.dav``` tests, so it seems it should be merged.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @dreid|
|-|-|

David, could you have a quick look to see if it's correct? I was wondering if we should not use id() instead of hashing the element.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @wsanchez|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez set owner to @therve|
|-|-|

Reviewed; looks good, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `closed`|
|-|-|

(In [22305]) Merge dav-hash-2960

Author: therve
Reviewers: wsanchez, exarkun
Fixes [#2960](https://github.com/twisted/twisted/issues/2960)

Add a hash method to twisted.web2.dav.element.base.WebDAVEmptyElement,
because instances where used as dictionary keys even though they were
not hashable. The behavior appeared recently with Python trunk, which used to 
allow new-style classes to have a default hash implementation even when
they override __eq__.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
